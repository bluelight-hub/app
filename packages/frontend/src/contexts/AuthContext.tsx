import React, { createContext, ReactNode, useEffect, useState } from 'react';

interface User {
    id: string;
    name: string;
    email: string;
    role: string;
}

export interface AuthContextType {
    user: User | null;
    isAuthenticated: boolean;
    isLoading: boolean;
    login: (email: string, password: string) => Promise<boolean>;
    logout: () => void;
}

// Mock-User für Entwicklungszwecke
const MOCK_USER: User = {
    id: '1',
    name: 'Test User',
    email: 'test@example.com',
    role: 'admin',
};

export const AuthContext = createContext<AuthContextType | null>(null);

interface AuthProviderProps {
    children: ReactNode;
}

export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
    const [user, setUser] = useState<User | null>(null);
    const [isLoading, setIsLoading] = useState(true);

    useEffect(() => {
        // Simuliere das Laden des Benutzers beim ersten Rendern
        const token = localStorage.getItem('auth_token');
        if (token) {
            // In einer realen Anwendung würden wir hier den Token verifizieren
            setUser(MOCK_USER);
        }
        setIsLoading(false);
    }, []);

    const login = async (email: string, password: string): Promise<boolean> => {
        setIsLoading(true);

        // Simuliere API-Aufruf mit Verzögerung
        return new Promise<boolean>((resolve) => {
            setTimeout(() => {
                // Mock-Authentifizierung - in der Produktion würde hier ein echter API-Aufruf stehen
                if (email === 'test@example.com' && password === 'password') {
                    setUser(MOCK_USER);
                    localStorage.setItem('auth_token', 'mock_jwt_token');
                    setIsLoading(false);
                    resolve(true);
                } else {
                    setIsLoading(false);
                    resolve(false);
                }
            }, 1000);
        });
    };

    const logout = () => {
        setUser(null);
        localStorage.removeItem('auth_token');
    };

    return (
        <AuthContext.Provider
            value={{
                user,
                isAuthenticated: !!user,
                isLoading,
                login,
                logout,
            }}
        >
            {children}
        </AuthContext.Provider>
    );
};

export default AuthContext; 