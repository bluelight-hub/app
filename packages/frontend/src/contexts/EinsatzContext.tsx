import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { Einsatz } from '@bluelight-hub/shared/client';
import { logger } from '../utils/logger';

interface EinsatzContextType {
    selectedEinsatz: Einsatz | null;
    selectEinsatz: (einsatz: Einsatz) => void;
    clearSelectedEinsatz: () => void;
    isEinsatzSelected: boolean;
}

const EinsatzContext = createContext<EinsatzContextType | null>(null);

interface EinsatzProviderProps {
    children: ReactNode;
}

export const EinsatzProvider: React.FC<EinsatzProviderProps> = ({ children }) => {
    const [selectedEinsatz, setSelectedEinsatz] = useState<Einsatz | null>(null);

    // Lade den gespeicherten Einsatz beim Start
    useEffect(() => {
        const savedEinsatzId = localStorage.getItem('selectedEinsatzId');
        const savedEinsatzName = localStorage.getItem('selectedEinsatzName');
        
        if (savedEinsatzId && savedEinsatzName) {
            // Erstelle ein minimales Einsatz-Objekt aus den gespeicherten Daten
            // In einer echten App würde man hier den vollständigen Einsatz laden
            setSelectedEinsatz({
                id: savedEinsatzId,
                name: savedEinsatzName,
                beschreibung: '',
                createdAt: new Date(),
                updatedAt: new Date()
            } as Einsatz);
            
            logger.debug('Loaded selected Einsatz from localStorage', { 
                id: savedEinsatzId, 
                name: savedEinsatzName 
            });
        }
    }, []);

    const selectEinsatz = (einsatz: Einsatz) => {
        setSelectedEinsatz(einsatz);
        localStorage.setItem('selectedEinsatzId', einsatz.id);
        localStorage.setItem('selectedEinsatzName', einsatz.name);
        logger.info('Einsatz selected', { id: einsatz.id, name: einsatz.name });
    };

    const clearSelectedEinsatz = () => {
        setSelectedEinsatz(null);
        localStorage.removeItem('selectedEinsatzId');
        localStorage.removeItem('selectedEinsatzName');
        logger.info('Selected Einsatz cleared');
    };

    return (
        <EinsatzContext.Provider
            value={{
                selectedEinsatz,
                selectEinsatz,
                clearSelectedEinsatz,
                isEinsatzSelected: !!selectedEinsatz
            }}
        >
            {children}
        </EinsatzContext.Provider>
    );
};

export const useEinsatzContext = () => {
    const context = useContext(EinsatzContext);
    if (!context) {
        throw new Error('useEinsatzContext must be used within an EinsatzProvider');
    }
    return context;
};

export default EinsatzContext;