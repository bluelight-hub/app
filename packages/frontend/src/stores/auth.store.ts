import { Store } from '@tanstack/react-store';
import type { UserResponseDto } from '@bluelight-hub/shared/client';

/**
 * Interface für den Authentifizierungs-Store
 */
export interface AuthState {
  /** Der aktuell authentifizierte Benutzer oder null wenn nicht angemeldet */
  user: UserResponseDto | null;
  /** Ob der Benutzer authentifiziert ist */
  isAuthenticated: boolean;
  /** Ob gerade eine Authentifizierung läuft */
  isLoading: boolean;
  /** Ob der Benutzer als Admin authentifiziert ist */
  isAdminAuthenticated: boolean;
}

/**
 * TanStack Store für die Authentifizierung
 *
 * Verwaltet den globalen Authentifizierungszustand der Anwendung.
 * Der Store ist framework-agnostisch und bietet fine-grained reactivity,
 * sodass Komponenten nur bei Änderungen der abonnierten Properties re-rendern.
 */
export const authStore = new Store<AuthState>({
  user: null,
  isAuthenticated: false,
  isLoading: false,
  isAdminAuthenticated: false,
});

/**
 * Actions für den Auth Store
 *
 * Kapselt alle Authentifizierungs-bezogenen Aktionen
 */
export const authActions = {
  /**
   * Setzt den authentifizierten Benutzer
   *
   * @param user - Die Benutzerdaten oder null zum Ausloggen
   */
  setUser: (user: UserResponseDto | null) => {
    authStore.setState((state) => ({
      ...state,
      user,
      isAuthenticated: !!user,
    }));
  },

  /**
   * Setzt den Loading-Status
   *
   * @param isLoading - Der neue Loading-Status
   */
  setLoading: (isLoading: boolean) => {
    authStore.setState((state) => ({
      ...state,
      isLoading,
    }));
  },

  /**
   * Meldet den Benutzer ab und löscht den Store
   */
  logout: () => {
    authStore.setState(() => ({
      user: null,
      isAuthenticated: false,
      isLoading: false,
      isAdminAuthenticated: false,
    }));
  },

  /**
   * Hilfsfunktion für Login/Registrierung
   * Setzt User und beendet Loading-Status
   *
   * @param user - Die Benutzerdaten nach erfolgreicher Authentifizierung
   */
  loginSuccess: (user: UserResponseDto) => {
    authStore.setState((state) => ({
      ...state,
      user,
      isAuthenticated: true,
      isLoading: false,
    }));
  },

  /**
   * Setzt den Admin-Authentifizierungsstatus
   *
   * @param isAdminAuthenticated - Ob der Benutzer als Admin authentifiziert ist
   */
  setAdminAuth: (isAdminAuthenticated: boolean) => {
    authStore.setState((state) => ({
      ...state,
      isAdminAuthenticated,
    }));
  },
};
