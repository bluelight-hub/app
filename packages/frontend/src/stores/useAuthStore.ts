import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';
import { api } from '../api';
import { logger } from '../utils/logger';
import { authStorage } from '../utils/authStorage';
import { AuthUserDto } from '@bluelight-hub/shared/client';

interface AuthState {
  user: AuthUserDto | null;
  isAuthenticated: boolean;
  isLoading: boolean;
  isRefreshing: boolean;
  accessToken: string | null;
  refreshToken: string | null;
  sessionExpiresAt: Date | null;
  lastRefreshAt: Date | null;
  refreshTimer: NodeJS.Timeout | null;
}

interface AuthActions {
  setUser: (user: AuthUserDto | null) => void;
  setTokens: (accessToken: string, refreshToken?: string, expiresIn?: number) => Promise<void>;
  login: (email: string, password: string) => Promise<{ success: boolean; error?: string }>;
  logout: () => Promise<void>;
  refreshAccessToken: () => Promise<boolean>;
  checkAuthStatus: () => Promise<void>;
  hasRole: (role: string) => boolean;
  hasPermission: (permission: string) => boolean;
  isAdmin: () => boolean;
  scheduleTokenRefresh: () => void;
  clearRefreshTimer: () => void;
}

export interface AuthStore extends AuthState, AuthActions {}

const REFRESH_BUFFER_SECONDS = 60; // Refresh 1 minute before expiry
const MIN_REFRESH_INTERVAL = 30000; // Minimum 30 seconds between refresh attempts

export const useAuthStore = create<AuthStore>()(
  devtools(
    persist(
      (set, get) => ({
        // Initial state
        user: null,
        isAuthenticated: false,
        isLoading: false,
        isRefreshing: false,
        accessToken: null,
        refreshToken: null,
        sessionExpiresAt: null,
        lastRefreshAt: null,
        refreshTimer: null,

        // Actions
        setUser: (user) => set({ user, isAuthenticated: !!user }),

        setTokens: async (accessToken, refreshToken, expiresIn) => {
          // Clear existing refresh timer
          get().clearRefreshTimer();

          // Calculate session expiry
          const sessionExpiresAt = expiresIn
            ? new Date(Date.now() + expiresIn * 1000)
            : new Date(Date.now() + 15 * 60 * 1000); // Default 15 minutes

          // Store tokens
          await authStorage.setTokens(accessToken, refreshToken);

          set({
            accessToken,
            refreshToken: refreshToken || get().refreshToken,
            sessionExpiresAt,
            lastRefreshAt: new Date(),
          });

          // Schedule next refresh
          get().scheduleTokenRefresh();

          logger.debug('Tokens updated', {
            expiresAt: sessionExpiresAt.toISOString(),
            hasRefreshToken: !!refreshToken,
          });
        },

        login: async (email, password) => {
          set({ isLoading: true });

          try {
            const response = await api.auth.authControllerLoginV1Raw({
              loginDto: { email, password },
            });
            const data = await response.value();

            if (data) {
              logger.info('Login successful', { email });

              // Set user and tokens
              if (data.user) {
                set({ user: data.user, isAuthenticated: true });
              }

              if (data.accessToken) {
                await get().setTokens(data.accessToken, data.refreshToken, data.expiresIn);
              }

              set({ isLoading: false });
              return { success: true };
            } else {
              set({ isLoading: false });
              return {
                success: false,
                error: 'No data received from server',
              };
            }
          } catch (error) {
            logger.error('Login error', { error });
            set({ isLoading: false });
            return {
              success: false,
              error: error instanceof Error ? error.message : 'Network error',
            };
          }
        },

        logout: async () => {
          try {
            // Clear refresh timer first
            get().clearRefreshTimer();

            // Call logout endpoint to revoke session
            await api.auth.authControllerLogoutV1();
            logger.info('Logout successful');
          } catch (error) {
            logger.error('Logout error', { error });
          } finally {
            // Clear local state and storage
            set({
              user: null,
              isAuthenticated: false,
              accessToken: null,
              refreshToken: null,
              sessionExpiresAt: null,
              lastRefreshAt: null,
            });
            await authStorage.clearTokens();
          }
        },

        refreshAccessToken: async () => {
          const state = get();

          // Prevent concurrent refresh attempts
          if (state.isRefreshing) {
            logger.debug('Refresh already in progress');
            return false;
          }

          // Check minimum interval between refresh attempts
          if (state.lastRefreshAt) {
            const timeSinceLastRefresh = Date.now() - state.lastRefreshAt.getTime();
            if (timeSinceLastRefresh < MIN_REFRESH_INTERVAL) {
              logger.debug('Skipping refresh - too soon since last attempt', {
                timeSinceLastRefresh,
              });
              return false;
            }
          }

          const { refreshToken } = state;
          if (!refreshToken) {
            logger.warn('No refresh token available');
            return false;
          }

          set({ isRefreshing: true });

          try {
            const response = await api.auth.authControllerRefreshTokenV1Raw({
              refreshTokenDto: { refreshToken },
            });

            const data = await response.value();
            if (data) {
              logger.info('Token refresh successful');

              // Update tokens and user data
              if (data.accessToken) {
                await get().setTokens(data.accessToken, data.refreshToken, data.expiresIn);
              }

              if (data.user) {
                set({ user: data.user });
              }

              set({ isRefreshing: false });
              return true;
            } else {
              set({ isRefreshing: false });
              return false;
            }
          } catch (error: any) {
            logger.error('Token refresh error', { error });

            // If refresh fails with 401, clear auth state
            if (error?.response?.status === 401 || error?.status === 401) {
              await get().logout();
            }

            set({ isRefreshing: false });
            return false;
          }
        },

        checkAuthStatus: async () => {
          set({ isLoading: true });

          try {
            const hasTokens = await authStorage.hasValidTokens();
            if (!hasTokens) {
              set({ isLoading: false });
              return;
            }

            // Load tokens from storage
            const tokens = await authStorage.getTokens();
            if (tokens.authToken) {
              set({
                accessToken: tokens.authToken,
                refreshToken: tokens.refreshToken || null,
              });
            }

            // Try to get current user
            try {
              const response = await api.auth.authControllerGetCurrentUserV1Raw();
              const userData = await response.value();
              if (userData) {
                set({
                  user: userData,
                  isAuthenticated: true,
                });
                logger.debug('User data loaded successfully');

                // Schedule token refresh
                get().scheduleTokenRefresh();
              } else {
                logger.warn('Failed to fetch user data');
                await get().logout();
              }
            } catch (error) {
              logger.error('Error fetching user data:', error);
              // Try to refresh token
              const refreshed = await get().refreshAccessToken();
              if (!refreshed) {
                await get().logout();
              }
            }
          } catch (error) {
            logger.error('Error checking auth status:', error);
          } finally {
            set({ isLoading: false });
          }
        },

        hasRole: (role) => {
          const { user } = get();
          return user?.roles?.includes(role) || false;
        },

        hasPermission: (permission) => {
          const { user } = get();
          return user?.permissions?.includes(permission) || false;
        },

        isAdmin: () => {
          const { user } = get();
          return (
            user?.roles?.some((role) => ['SUPER_ADMIN', 'ADMIN', 'SUPPORT'].includes(role)) || false
          );
        },

        scheduleTokenRefresh: () => {
          const state = get();

          // Clear any existing timer
          get().clearRefreshTimer();

          if (!state.sessionExpiresAt || !state.refreshToken) {
            logger.debug('Cannot schedule refresh - missing expiry or refresh token');
            return;
          }

          const now = Date.now();
          const expiresAt = state.sessionExpiresAt.getTime();
          const refreshAt = expiresAt - REFRESH_BUFFER_SECONDS * 1000;
          const delay = Math.max(refreshAt - now, 0);

          if (delay > 0) {
            logger.debug('Scheduling token refresh', {
              expiresAt: state.sessionExpiresAt.toISOString(),
              refreshIn: Math.round(delay / 1000) + 's',
            });

            const timer = setTimeout(async () => {
              logger.debug('Executing scheduled token refresh');
              await get().refreshAccessToken();
            }, delay);

            set({ refreshTimer: timer });
          } else {
            // Token already expired or about to expire, refresh immediately
            logger.debug('Token expired or about to expire, refreshing immediately');
            get().refreshAccessToken();
          }
        },

        clearRefreshTimer: () => {
          const { refreshTimer } = get();
          if (refreshTimer) {
            clearTimeout(refreshTimer);
            set({ refreshTimer: null });
          }
        },
      }),
      {
        name: 'auth-storage',
        partialize: (state) => ({
          // Only persist non-sensitive data
          user: state.user,
          isAuthenticated: state.isAuthenticated,
        }),
      },
    ),
    {
      name: 'AuthStore',
    },
  ),
);

// Selector hooks for common use cases
export const useUser = () => useAuthStore((state) => state.user);
export const useIsAuthenticated = () => useAuthStore((state) => state.isAuthenticated);
export const useIsLoading = () => useAuthStore((state) => state.isLoading);
export const useAuthActions = () =>
  useAuthStore((state) => ({
    login: state.login,
    logout: state.logout,
    hasRole: state.hasRole,
    hasPermission: state.hasPermission,
    isAdmin: state.isAdmin,
  }));
