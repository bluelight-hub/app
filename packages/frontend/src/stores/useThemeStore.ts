import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';

interface ThemeState {
    dark: boolean;
    auto: boolean;
    // Für manuelle Änderungen durch den Benutzer
    setManualDark: (dark: boolean) => void;
    // Für automatische Änderungen durch das System
    setSystemDark: (dark: boolean) => void;
    setAuto: (auto: boolean) => void;
}

// Get initial system dark mode preference
const getInitialDarkMode = () => {
    if (typeof window === 'undefined') return false;
    return window.matchMedia('(prefers-color-scheme: dark)').matches;
};

export const useThemeStore = create<ThemeState>()(
    devtools(
        persist(
            (set) => {
                // Initialize with system preference
                const systemDark = getInitialDarkMode();

                return {
                    dark: systemDark,
                    auto: true,
                    // Manuelle Änderung durch den Benutzer
                    setManualDark: (dark) =>
                        set(() => ({
                            dark,
                            auto: false, // Automatik wird ausgeschaltet
                        })),
                    // Automatische Änderung durch das System
                    setSystemDark: (dark) =>
                        set((state) => ({
                            dark: state.auto ? dark : state.dark, // Nur ändern wenn auto aktiv
                            auto: state.auto, // Auto-Status bleibt unverändert
                        })),
                    setAuto: (auto) =>
                        set((state) => ({
                            auto,
                            dark: auto ? getInitialDarkMode() : state.dark,
                        })),
                };
            },
            {
                name: 'theme-storage',
            }
        )
    )
); 