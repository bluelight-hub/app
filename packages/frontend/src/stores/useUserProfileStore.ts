import { create } from 'zustand';
import { devtools } from 'zustand/middleware';

interface UserProfile {
    name: string;
    imageUrl: string;
    email?: string;
}

interface UserProfileState {
    profile: UserProfile | null;
    isLoading: boolean;
    error: string | null;
    setProfile: (profile: UserProfile) => void;
    updateProfile: (updates: Partial<UserProfile>) => void;
    clearProfile: () => void;
}

export const useUserProfileStore = create<UserProfileState>()(
    devtools(
        (set) => ({
            profile: {
                name: 'Robert Retter',
                imageUrl: 'https://randomuser.me/api/portraits/women/11.jpg',
            },
            isLoading: false,
            error: null,
            setProfile: (profile) => set({ profile, error: null }),
            updateProfile: (updates) =>
                set((state) => ({
                    profile: state.profile ? { ...state.profile, ...updates } : null,
                    error: null,
                })),
            clearProfile: () => set({ profile: null, error: null }),
        }),
        { name: 'user-profile-store' }
    )
); 