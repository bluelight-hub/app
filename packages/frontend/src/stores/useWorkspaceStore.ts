import { create, StateCreator } from 'zustand';
import { devtools } from 'zustand/middleware';

export interface WorkspaceTab {
    id: string;
    title: string;
    type: 'empty' | 'project' | 'settings';
    active: boolean;
}

interface WorkspaceState {
    tabs: WorkspaceTab[];
    activeTabId: string | null;
    addTab: (type: WorkspaceTab['type'], title?: string) => void;
    removeTab: (id: string) => void;
    setActiveTab: (id: string) => void;
}

type WorkspaceStore = StateCreator<
    WorkspaceState,
    [["zustand/devtools", never]],
    [],
    WorkspaceState
>;

export const useWorkspaceStore = create<WorkspaceState>()(
    devtools(
        ((set) => ({
            tabs: [],
            activeTabId: null,
            addTab: (type: WorkspaceTab['type'], title?: string) => set((state: WorkspaceState) => {
                const newTab: WorkspaceTab = {
                    id: crypto.randomUUID(),
                    title: title || `New ${type} ${state.tabs.length + 1}`,
                    type,
                    active: true,
                };

                return {
                    tabs: [...state.tabs.map(tab => ({ ...tab, active: false })), newTab],
                    activeTabId: newTab.id,
                };
            }),
            removeTab: (id: string) => set((state: WorkspaceState) => {
                const newTabs = state.tabs.filter(tab => tab.id !== id);
                const newActiveTabId = state.activeTabId === id
                    ? newTabs[newTabs.length - 1]?.id || null
                    : state.activeTabId;

                return {
                    tabs: newTabs,
                    activeTabId: newActiveTabId,
                };
            }),
            setActiveTab: (id: string) => set((state: WorkspaceState) => ({
                tabs: state.tabs.map(tab => ({
                    ...tab,
                    active: tab.id === id,
                })),
                activeTabId: id,
            })),
        })) as WorkspaceStore,
        { name: 'workspace-store' }
    )
); 