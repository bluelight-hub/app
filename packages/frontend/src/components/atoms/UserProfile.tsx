import { Dropdown, MenuProps } from 'antd';
import { PiSignOut, PiUser } from 'react-icons/pi';
import React, { useState } from 'react';
import { useEinsatzContext } from '../../contexts/EinsatzContext';
import { useUserProfileStore } from '../../stores/useUserProfileStore';
import { logout } from '../../utils/auth';
import { BaseAtomProps } from '../../utils/types';

interface UserProfileProps extends BaseAtomProps {
    href: string;
    onClick?: () => void;
    hideText?: boolean;
}

/**
 * User profile component for the sidebar
 */
const UserProfile: React.FC<UserProfileProps> = ({
    href,
    onClick,
    hideText = false,
    'data-testid': dataTestId = 'user-profile'
}) => {
    const profile = useUserProfileStore((state) => state.profile);
    const { clearSelectedEinsatz } = useEinsatzContext();
    const [loading, setLoading] = useState(false);

    const handleLogout = async () => {
        setLoading(true);
        try {
            await logout(clearSelectedEinsatz);
        } catch (error) {
            // Fehler wird bereits in der logout-Funktion behandelt
        } finally {
            setLoading(false);
        }
    };

    const items: MenuProps['items'] = [
        {
            key: 'profile',
            icon: <PiUser />,
            label: 'Profil',
            onClick: () => {
                if (onClick) onClick();
            }
        },
        {
            type: 'divider',
        },
        {
            key: 'logout',
            icon: <PiSignOut />,
            label: 'Ausloggen',
            danger: true,
            onClick: handleLogout,
            disabled: loading
        },
    ];

    if (!profile) return null;

    return (
        <Dropdown menu={{ items }} trigger={['click']} placement="topRight">
            <a
                href="#"
                onClick={(e) => e.preventDefault()}
                data-testid={dataTestId}
                className="flex w-full h-12 items-center gap-x-4 px-6 py-2 text-sm/6 font-semibold text-gray-900 dark:text-gray-100 hover:bg-gray-50 dark:hover:bg-gray-800 cursor-pointer"
            >
                <img
                    className="size-8 rounded-full bg-gray-50 dark:bg-gray-800"
                    src={profile.imageUrl}
                    alt={profile.name}
                />
                <span className="sr-only">Dein Profil</span>
                {!hideText && <span aria-hidden="true">{profile.name}</span>}
            </a>
        </Dropdown>
    );
};

export default UserProfile;