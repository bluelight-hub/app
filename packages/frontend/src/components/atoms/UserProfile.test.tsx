import {act, fireEvent, render, screen} from '@testing-library/react';
import {describe, expect, it, vi} from 'vitest';
import {EinsatzProvider} from '../../contexts/EinsatzContext';
import {useUserProfileStore} from '../../stores/useUserProfileStore';
import UserProfile from './UserProfile';

// Definiere die benötigten Typen lokal, da sie nicht exportiert werden
interface UserProfile {
    name: string;
    imageUrl: string;
    email?: string;
}

interface MockUserProfileState {
    profile: UserProfile | null;
    isLoading: boolean;
    error: string | null;
    setProfile: (profile: UserProfile) => void;
    updateProfile: (updates: Partial<UserProfile>) => void;
    clearProfile: () => void;
}

// Mock für den Store mit verschiedenen Implementierungen
const mockProfile = {
    name: 'Test User',
    imageUrl: 'test-image-url',
    email: 'test@example.com'
};

// Mock-Funktionen für den Store
const mockSetProfile = vi.fn();
const mockUpdateProfile = vi.fn();
const mockClearProfile = vi.fn();

// Mock für useUserProfileStore
vi.mock('../../stores/useUserProfileStore', () => ({
    useUserProfileStore: vi.fn((selector) => selector({
        profile: mockProfile,
        isLoading: false,
        error: null,
        setProfile: mockSetProfile,
        updateProfile: mockUpdateProfile,
        clearProfile: mockClearProfile
    }))
}));

// Mock für auth utils
vi.mock('../../utils/auth', () => ({
    logout: vi.fn()
}));

// Helper function to render with EinsatzProvider
const renderWithEinsatzProvider = (component: React.ReactElement) => {
    return render(<EinsatzProvider>{component}</EinsatzProvider>);
};

describe('UserProfile', () => {
    // Unit Test - Grundlegendes Rendering
    it('should render with profile data', () => {
        renderWithEinsatzProvider(<UserProfile href="/profile" data-testid="user-profile" />);

        const profileElement = screen.getByTestId('user-profile');
        expect(profileElement).toBeInTheDocument();
        expect(screen.getByText('Test User')).toBeInTheDocument();
    });

    // Unit Test - Prüft, ob das Bild korrekt gerendert wird
    it('should render the profile image correctly', () => {
        renderWithEinsatzProvider(<UserProfile href="/profile" data-testid="user-profile" />);

        const imageElement = screen.getByAltText('Test User');
        expect(imageElement).toBeInTheDocument();
        expect(imageElement).toHaveAttribute('src', 'test-image-url');
        expect(imageElement).toHaveClass('rounded-full');
    });

    // Unit Test - Prüft, ob onClick richtig aufgerufen wird (jetzt Dropdown)
    it('should handle onClick event', async () => {
        const handleClick = vi.fn();
        renderWithEinsatzProvider(<UserProfile href="/profile" onClick={handleClick} data-testid="user-profile" />);

        // Dropdown trigger klicken
        await act(async () => {
            fireEvent.click(screen.getByTestId('user-profile'));
        });
        
        // Das onclick wird nicht mehr direkt aufgerufen, da wir ein Dropdown haben
        // Stattdessen testen wir, dass das Dropdown funktioniert
        expect(screen.getByTestId('user-profile')).toBeInTheDocument();
    });

    // Unit Test - Prüft, ob der onClick Handler im Profil-MenuItem funktioniert
    it('should call onClick when profile menu item is clicked', async () => {
        const handleClick = vi.fn();
        renderWithEinsatzProvider(<UserProfile href="/profile" onClick={handleClick} data-testid="user-profile" />);

        // Dropdown öffnen
        await act(async () => {
            fireEvent.click(screen.getByTestId('user-profile'));
        });

        // Auf "Profil" klicken
        const profileMenuItem = screen.getByText('Profil');
        await act(async () => {
            fireEvent.click(profileMenuItem);
        });

        expect(handleClick).toHaveBeenCalledTimes(1);
    });

    // Unit Test - Prüft, ob der hideText Parameter funktioniert
    it('should hide text when hideText is true', () => {
        renderWithEinsatzProvider(<UserProfile href="/profile" hideText={true} data-testid="user-profile" />);

        // Bild sollte noch sichtbar sein
        expect(screen.getByAltText('Test User')).toBeInTheDocument();

        // Name sollte nicht sichtbar sein
        expect(screen.queryByText('Test User')).not.toBeInTheDocument();
    });

    // Integration Test - Im Layout-Kontext
    it('should work properly in a sidebar layout', () => {
        render(
            <EinsatzProvider>
                <div data-testid="sidebar">
                    <nav>
                        <UserProfile href="/profile" data-testid="user-profile" />
                    </nav>
                </div>
            </EinsatzProvider>
        );

        const sidebar = screen.getByTestId('sidebar');
        const profileElement = screen.getByTestId('user-profile');
        expect(sidebar).toContainElement(profileElement);
    });

    // Snapshot Test
    it('should match snapshot', () => {
        const { container } = renderWithEinsatzProvider(<UserProfile href="/profile" data-testid="user-profile" />);
        expect(container).toMatchSnapshot();
    });

    // Snapshot Test with hideText
    it('should match snapshot with hidden text', () => {
        const { container } = renderWithEinsatzProvider(<UserProfile href="/profile" hideText={true} data-testid="user-profile" />);
        expect(container).toMatchSnapshot();
    });

    // Test Fall: Kein Profil vorhanden
    it('should render nothing when profile is not available', () => {
        // Temporärer Mock-Override für diesen Test
        vi.mocked(useUserProfileStore).mockImplementationOnce((selector) => selector({
            profile: null,
            isLoading: false,
            error: null,
            setProfile: mockSetProfile,
            updateProfile: mockUpdateProfile,
            clearProfile: mockClearProfile
        } as MockUserProfileState));

        const { container } = renderWithEinsatzProvider(<UserProfile href="/profile" data-testid="user-profile" />);
        expect(container.firstChild).toBeNull();
    });
}); 