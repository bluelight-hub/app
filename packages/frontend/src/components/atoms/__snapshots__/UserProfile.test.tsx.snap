// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserProfile > should match snapshot 1`] = `
<div>
  <div
    class="px-6 py-2"
  >
    <button
      class="ant-btn css-dev-only-do-not-override-ut69n1 ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-dropdown-trigger w-full"
      data-testid="user-profile"
      type="button"
    >
      <span
        class="ant-btn-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M230.92,212c-15.23-26.33-38.7-45.21-66.09-54.16a72,72,0,1,0-73.66,0C63.78,166.78,40.31,185.66,25.08,212a8,8,0,1,0,13.85,8c18.84-32.56,52.14-52,89.07-52s70.23,19.44,89.07,52a8,8,0,1,0,13.85-8ZM72,96a56,56,0,1,1,56,56A56.06,56.06,0,0,1,72,96Z"
          />
        </svg>
      </span>
      <span>
        Test User
      </span>
    </button>
  </div>
</div>
`;

exports[`UserProfile > should match snapshot with hidden text 1`] = `
<div>
  <div
    class=""
  >
    <button
      class="ant-btn css-dev-only-do-not-override-ut69n1 ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-icon-only ant-dropdown-trigger w-full"
      data-testid="user-profile"
      type="button"
    >
      <span
        class="ant-btn-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M230.92,212c-15.23-26.33-38.7-45.21-66.09-54.16a72,72,0,1,0-73.66,0C63.78,166.78,40.31,185.66,25.08,212a8,8,0,1,0,13.85,8c18.84-32.56,52.14-52,89.07-52s70.23,19.44,89.07,52a8,8,0,1,0,13.85-8ZM72,96a56,56,0,1,1,56,56A56.06,56.06,0,0,1,72,96Z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
